### PR TITLE
611 bug fix baseline traceroute

### PR DESCRIPTION
**Pull request by CAfitech**
Bugs/Issues:

- Interfaces did not have appropriate docstring.
- The FormValuesType interface was located inside the main component.
- Every function did not have appropriate docstring.
- Every React hook did not have an appropriate inline comment.
- Every component constants did not have appropriate inline comments.
- The title consisted of more than the tools name.
- The description was overly lengthy and wasn't just from the source documentation.
- The steps were included in the description instead of separately.
-  The steps were not included or formatted.
- The sourceLink was not included.
- The tutorial was not included in the code.
- The dependencies were not included.
- The form did not have clear defined comments.
-  useEffect was not incorporated in the code.
- All of the dependencies components were not installed.
- Handle Process Data was not incorporated in the code.
- Handle Process Termination was not incorporated in the code.
- The onSubmit function did not have clear defined comments and docstring
- The clearOutput was not clearly defined.
- The outputV2 was not defined neither incorporated in the code.
- The Render component was not incorporated.

Fixes:

- Added all Interfaces with appropriate docstring.
- Moved the FormValuesType interface so its located outside the main component.
- Modified all the functions to have appropriate docstring.
- Modified so that all React hook have appropriate in line comments.
- Modified so that all component constants have appropriate in line comments.
- Modified the title to only include the name of the tool.
- Modified the description to be concise and accurate as per the checklist.
- Modified the code by adding the steps which are those only applicable to the feature.
- Modified the steps to properly format them.
- Modified to add the sourceLink containing the link to the source component.
- Modified to add the tutorial with an empty string.
- Modified to add the dependencies with the required packages for the tool.
- Modified so that the form is clearly defined including comments.
- Modified to incorporate the useEffect, clearly defining including comments.
- Modified to install the dependencies required.
- Modified to incorporate Handle Process Data with clearly defined docstring and comments
- Modified to incorporate Handle Process Termination with clearly defined docstring and comments
- Modified the onSubmit function to include clearly defined docstrings and comments
- Modified to incorporate the outputV2 clearly defined and in use.
- Modified to include the Render Component within the return of the component function.

### CODE REVIEW

File: src/components/Traceroute/Traceroute.tsx [hunk 1]
```diff
@@ -1,52 +1,40 @@
 import { Button, NativeSelect, Stack, TextInput } from "@mantine/core";
 import { useForm } from "@mantine/form";
-import { useCallback, useState } from "react";
+import { useCallback, useEffect, useState } from "react";
 import { CommandHelper } from "../../utils/CommandHelper";
 import ConsoleWrapper from "../ConsoleWrapper/ConsoleWrapper";
 import { UserGuide } from "../UserGuide/UserGuide";
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:6]: You still need to upgrade the component to use RenderComponent as userguide is deprecated. 
```suggestion
import { RenderComponent } from "../UserGuide/UserGuide";
```

File: src/components/Traceroute/Traceroute.tsx [hunk 2]
```diff
@@ -1,52 +1,40 @@
...
+import InstallationModal from "../InstallationModal/InstallationModal";
+import { RenderComponent } from "../UserGuide/UserGuide";
+
+/**
+ * FormValuesType defines the structure for the form values used in the TracerouteTool component.
+ * @field hostname: The hostname or IP address for the traceroute operation.
+ * @field portNumber:The port number to be used, currently not utilized in the traceroute operations.
+ * @field tracerouteSwitch: The selected type of traceroute scan.
+ * @field tracerouteOptions: Custom traceroute options provided by the user.
+ */
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:18]: Good work here! You put some effort to list the fields and is noted. Good work.

File: src/components/Traceroute/Traceroute.tsx [hunk 3]
```diff
@@ -1,52 +1,40 @@
...
 interface FormValuesType {
     hostname: string;
     portNumber: string;
     tracerouteSwitch: string;
     tracerouteOptions: string;
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:23]: Need to use camelCase
```suggestion
    traceRouteSwitch: string;
    traceRouteOptions: string;
```

File: src/components/Traceroute/Traceroute.tsx [hunk 4]
```diff
@@ -1,52 +1,40 @@
...
-//Traceroute Options
-const tracerouteSwitch = [
-    "Traceroute ICMP scan",
-    "Traceroute TCP scan",
-    "Traceroute UDP scan",
-    "Traceroute custom scan",
-];
-
-//Tool name must be capital or jsx will cry out errors :P
+// Component State Variables.
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:26]: This needs to be inside the component function.

File: src/components/Traceroute/Traceroute.tsx [hunk 5]
```diff
@@ -1,52 +1,40 @@
...
 const TracerouteTool = () => {
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:27]: Add docstring. 
```suggestion
/**
 * The Traceroute component.
 * @returns The Traceroute component.
 */
const TracerouteTool = () => {
```

Comment (bradleybogg) [reply src/components/Traceroute/Traceroute.tsx:27]: Rename component Traceroute
```suggestion
const Traceroute = () => {
```

File: src/components/Traceroute/Traceroute.tsx [hunk 6]
```diff
@@ -1,52 +1,40 @@
...
-    var [output, setOutput] = useState("");
-    const [selectedScanOption, setSelectedTracerouteOption] = useState("");
-
+    var [output, setOutput] = useState(""); //State to store the output from the traceroute command
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:28]: Change to be a constant
```suggestion
    const [output, setOutput] = useState(""); //State to store the output from the traceroute command
```

File: src/components/Traceroute/Traceroute.tsx [hunk 7]
```diff
@@ -1,52 +1,40 @@
...
+    const [loading, setLoading] = useState(false); // State variable to indicate loading state.
+    const [pid, setPid] = useState(""); // State variable to store the process ID of the command execution.
+    const [selectedScanOption, setSelectedTracerouteOption] = useState(""); // State to store the selected scan type.
+    const [isCommandAvailable, setIsCommandAvailable] = useState(false); // State variable to check if the command is available.
+    const [opened, setOpened] = useState(!isCommandAvailable); // State variable that indicates if the model is opened.
+    const [LoadingModal, setLoadingModal] = useState(true); // State variable to indicate loading state of the model.
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:34]: Where is this being used and why? This should be display over.

File: src/components/Traceroute/Traceroute.tsx [hunk 8]
```diff
@@ -1,52 +1,40 @@
...
-
-//Tool name must be capital or jsx will cry out errors :P
+// Component State Variables.
 const TracerouteTool = () => {
-    var [output, setOutput] = useState("");
-    const [selectedScanOption, setSelectedTracerouteOption] = useState("");
-
+    var [output, setOutput] = useState(""); //State to store the output from the traceroute command
+    const [loading, setLoading] = useState(false); // State variable to indicate loading state.
+    const [pid, setPid] = useState(""); // State variable to store the process ID of the command execution.
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:30]: This needs to be used in cancel function. Currently not in use

File: src/components/Traceroute/Traceroute.tsx [hunk 9]
```diff
@@ -1,52 +1,40 @@
...
-];
-
-//Tool name must be capital or jsx will cry out errors :P
+// Component State Variables.
 const TracerouteTool = () => {
-    var [output, setOutput] = useState("");
-    const [selectedScanOption, setSelectedTracerouteOption] = useState("");
-
+    var [output, setOutput] = useState(""); //State to store the output from the traceroute command
+    const [loading, setLoading] = useState(false); // State variable to indicate loading state.
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:29]: Loading Modal usage. Update in return function

File: src/components/Traceroute/Traceroute.tsx [hunk 10]
```diff
@@ -1,52 +1,40 @@
...
+    const [pid, setPid] = useState(""); // State variable to store the process ID of the command execution.
+    const [selectedScanOption, setSelectedTracerouteOption] = useState(""); // State to store the selected scan type.
+    const [isCommandAvailable, setIsCommandAvailable] = useState(false); // State variable to check if the command is available.
+    const [opened, setOpened] = useState(!isCommandAvailable); // State variable that indicates if the model is opened.
+    const [LoadingModal, setLoadingModal] = useState(true); // State variable to indicate loading state of the model.
+    const [allowSave, setAllowSave] = useState(false); // State variable indicating whether the current state is valid and the results can be saved.
+    const [hasSaved, setHasSaved] = useState(false); // State variable that tracks whether the results have already been saved to avoid redundant operations.
+    // Form hook to handle form input.
     let form = useForm({
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:38]: Doc String

Comment (bradleybogg) [reply src/components/Traceroute/Traceroute.tsx:38]: Update to camelCase

File: src/components/Traceroute/Traceroute.tsx [hunk 11]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
+        "The Traceroute tool provides a utility for displaying the route that IP packets have used as they travel to a particular network or host.";
+    const steps =
+        "Step 1: Enter a Hostname/IP address.\n" +
+        "Step 2: (Optional) Enter any additional options.\n" +
+        "Step 3: Select a scan option.\n" +
+        "Step 4: Click Scan to commence Traceroute operation.\n" +
+        "Step 5: View the Output block below to view the results of the tool's execution.";
+    const sourceLink = "https://www.kali.org/tools/"; //Link to the source code(or kali tools).
+    const tutorial = ""; //Link to the official documentation/tutorial.
+    const dependencies = ["traceroute"]; //Contains the dependencies required by the component.
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:58]: Move to before the form.

Comment (bradleybogg) [reply src/components/Traceroute/Traceroute.tsx:58]: Update comments to ensure they all have a space 

// This is a comment
//This is not a comment

Comment (bradleybogg) [reply src/components/Traceroute/Traceroute.tsx:58]: Ensure capitalisation is correct

Comment (bradleybogg) [reply src/components/Traceroute/Traceroute.tsx:58]: Update source link to actual or leave blank but not incorrect location

File: src/components/Traceroute/Traceroute.tsx [hunk 12]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
+
+    /**
+     * handleProcessTermination: Callback to handle the termination of the child process.
+     * Once the process termination is handled, it clears the process PID reference and
+     * deactivates the loading overlay.
+     * @param {object} param - An object containing information about the process termination.
+     * @param {number} param.code - The exit code of the terminated process.
+     * @param {number} param.signal - The signal code indicating how the process was terminated.
+     */
+
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:92]: Remove space

File: src/components/Traceroute/Traceroute.tsx [hunk 13]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
+        [handleProcessData] // Dependency on the handleProcessData callback
+    );
+
+    //Traceroute Options
+    const tracerouteSwitch = [
+        "Traceroute ICMP scan",
+        "Traceroute TCP scan",
+        "Traceroute UDP scan",
+        "Traceroute custom scan",
+    ]; // Options for different types of traceroute scans
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:123]: Remove trailing comment to before.
Move to component constants section.

Comment (bradleybogg) [reply src/components/Traceroute/Traceroute.tsx:123]: Change to camelCase.
Fix comment.
```suggestion
    // Traceroute Options used within the form.
    const traceRouteSwitch = [
        "Traceroute ICMP scan",
        "Traceroute TCP scan",
        "Traceroute UDP scan",
        "Traceroute custom scan",
    ];
```

File: src/components/Traceroute/Traceroute.tsx [hunk 14]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
+     */
     const onSubmit = async (values: FormValuesType) => {
         let args = [``];
 
-        //Switch case
+        // Switch case to handle different traceroute scan options based on user selection.
         switch (values.tracerouteSwitch) {
-            case "Traceroute ICMP scan": //traceroute syntax: traceroute -I <hostname>
+            case "Traceroute ICMP scan":
+                // Traceroute ICMP scan uses the '-I' option
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:138]: Move comments to before the case switch

File: src/components/Traceroute/Traceroute.tsx [hunk 15]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
                 } catch (e: any) {
-                    setOutput(e);
+                    setOutput(e); // Sets error to the state if command fails.
                 }
 
                 break;
 
-            case "Traceroute TCP scan": //traceroute syntax: traceroute -T <hostname>
+            case "Traceroute TCP scan":
+                // TRaceroute TCP scan uses the '-T' option.
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:155]: Move comments before

File: src/components/Traceroute/Traceroute.tsx [hunk 16]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
                     let output = await CommandHelper.runCommand("bash", args);
-                    setOutput(output);
+                    setOutput(output); //Sets the output to the state.
+                    setAllowSave(true);
                 } catch (e: any) {
-                    setOutput(e);
+                    setOutput(e); //Sets error to the state if command fails.
                 }
 
                 break;
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:168]: Review requirement for a break this is likely not required in switch stack.

File: src/components/Traceroute/Traceroute.tsx [hunk 17]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
                     let output = await CommandHelper.runCommand("bash", args);
-                    setOutput(output);
+                    setOutput(output); // Sets the output to the state.
+                    setAllowSave(true);
                 } catch (e: any) {
-                    setOutput(e);
+                    setOutput(e); // Sets errors to the state if command fails.
                 }
 
                 break;
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:184]: Review requirement for a break this is likely not required in switch stack.

File: src/components/Traceroute/Traceroute.tsx [hunk 18]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
                     let output = await CommandHelper.runCommand("bash", args);
-                    setOutput(output);
+                    setOutput(output); // Set the output to the state.
+                    setAllowSave(true);
                 } catch (e: any) {
-                    setOutput(e);
+                    setOutput(e); // Sets errors to the state if the command fails.
                 }
 
                 break;
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:200]: Review requirement for a break this is likely not required in switch stack.

File: src/components/Traceroute/Traceroute.tsx [hunk 19]
```diff
@@ -55,89 +43,206 @@ const TracerouteTool = () => {
...
+                        description={"Type of scan to perform"}
+                    />
+                    <Button type={"submit"}>start traceroute</Button>
+                    {SaveOutputToTextFile_v2(output, allowSave, hasSaved, handleSaveComplete)}
+                    <ConsoleWrapper output={output} clearOutputCallback={clearOutput} />
+                </Stack>
+            </form>
+        </RenderComponent>
     );
 };
```
Comment (bradleybogg) [inline src/components/Traceroute/Traceroute.tsx:248]: rectify such that tool is reflective only of the actual tool.

**Review by bradleybogg — CHANGES_REQUESTED** [review]
Good work @CAfitech 

Please see requested changes.
Ensure that you follow the guides given as some stuff here deviates.

Kind Regards 

Bradley

File: src/components/Traceroute/Traceroute.tsx (hunk 1 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: I have removed User guide and I have updated the code to use the Render Component.

File: src/components/Traceroute/Traceroute.tsx (hunk 4 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: I have moved this comment inside the component function

File: src/components/Traceroute/Traceroute.tsx (hunk 5 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: I have done both of those added the doc string and renamed the tool just Traceroute.

File: src/components/Traceroute/Traceroute.tsx (hunk 6 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: I changed the var [output, setOutput] to be constant.

File: src/components/Traceroute/Traceroute.tsx (hunk 9 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:32]: Updated the return function to use the learning modal.

File: src/components/Traceroute/Traceroute.tsx (hunk 8 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:33]: Had some issue implementing handleProcessTermination call back, need help in that.

File: src/components/Traceroute/Traceroute.tsx (hunk 7 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: Not sure what you are refering to here a bit confused.

File: src/components/Traceroute/Traceroute.tsx (hunk 10 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:41]: Not sure which one you are referring to here.

File: src/components/Traceroute/Traceroute.tsx (hunk 11 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:54]: Moved before the form, updated the comments to have proper spacing, ensure capitalization is correct, updated the source link for the actual link.

File: src/components/Traceroute/Traceroute.tsx (hunk 12 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:96]: I am not sure what is being referred to here .

File: src/components/Traceroute/Traceroute.tsx (hunk 13 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: Modified the code to do the above.

File: src/components/Traceroute/Traceroute.tsx (hunk 14 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx]: Moved the comments before the case switch

File: src/components/Traceroute/Traceroute.tsx (hunk 16 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:165]: I tested the break statement and I believe it is required as it negatively impacts the result and execution providing incorrect executions.

File: src/components/Traceroute/Traceroute.tsx (hunk 17 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:180]: I tested the break statement and I believe it is required as it negatively impacts the result and execution providing incorrect executions.

File: src/components/Traceroute/Traceroute.tsx (hunk 18 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:195]: I tested the break statement and I believe it is required as it negatively impacts the result and execution providing incorrect executions.

File: src/components/Traceroute/Traceroute.tsx (hunk 19 above)
Comment (CAfitech) [reply src/components/Traceroute/Traceroute.tsx:252]: Could you please explain that a bit more not sure how to implement this.

**Comment by CAfitech** [comment]
I have completed all the required tasks that were asked. Ready for review.

**Review by michaeljpigott — APPROVED** [review]
Thank you @CAfitech 

You have successfully baselined Traceroute. 

I am having an issue getting the tool to run, however, I am having the same issue on main, so I do not believe it relates to your baselining work. 

image

We may need to log an issue to ensure users have a copy of the shell file on their local computer. 

Kind regards 

Michael

**Pull request merged**